### PR TITLE
2269 update reports with exploitation events

### PR DESCRIPTION
**Pull request by cakekoa**
# What does this PR do?

Fixes part of #2269.

Updates the issues reports. This affects the "Security Findings" > "Immediate Threats" and "Potential Security Issues" lists, as well as the recommendations. Note that these recommendations are more vague/less useful now because we don't have the usernames or service/ports available to indicate which user or which service/port is vulnerable.

Other PRs will focus on:
- Removing PTHReportService
- Updating cross-segmentation reports to use the repositories
- Updating monkey_exploitation to use the repositories

## PR Checklist
* [x] Have you added an explanation of what your changes do and why you'd like to include them?
* [x] Is the TravisCI build passing?
* [ ] ~~Was the CHANGELOG.md updated to reflect the changes?~~
* [x] Was the documentation framework updated to reflect the changes?
* [x] Have you checked that you haven't introduced any duplicate code?

## Testing Checklist

* [ ] ~~Added relevant unit tests?~~
* [x] Have you successfully tested your changes locally? Elaborate:
    > Tested by running in the zoo envionment
* [x] If applicable, add screenshots or log transcripts of the feature working
    Here we see the map:
    Screen Shot 2022-10-11 at 9 08 33 AM

    Observe that the "Immediate Threats" are still valid:
    Screen Shot 2022-10-11 at 9 07 48 AM

    Observe that "Potential Security Issues" is valid. Also observe that the "Machine related recommendations" list is populated:
    Screen Shot 2022-10-11 at 9 28 48 AM

    Observe that the expanded items in "Machine related recommendations" are correct:
    Screen Shot 2022-10-11 at 9 30 10 AM
    Screen Shot 2022-10-11 at 9 30 46 AM

### CODE REVIEW

File: monkey/monkey_island/cc/ui/src/components/report-components/security/issues/SshIssue.js [hunk 1]
```diff
@@ -7,33 +7,30 @@ export function sshIssueOverview() {
...
-          The Monkey authenticated over the SSH protocol with user <span
-          className="badge badge-success">{issue.username}</span> and its password.
-        </CollapsibleWellComponent>
-      </>
-    );
+    <>
+      Change user passwords to a complex one-use password that is not shared with other computers on the network.
+      <CollapsibleWellComponent>
+        The machine <span className="badge badge-primary">{issue.machine}</span> (<span
+          className="badge badge-info" style={{ margin: '2px' }}>{issue.ip_address}</span>) is vulnerable to a <span
```
Comment (shreyamalviya) [inline monkey/monkey_island/cc/ui/src/components/report-components/security/issues/SshIssue.js:14]:
```suggestion
          className="badge badge-info" style={{ margin: '2px' }}>{issue.ip_address}</span>) is vulnerable to an <span
```

File: monkey/monkey_island/cc/ui/src/components/report-components/security/issues/SshIssue.js [hunk 2]
```diff
@@ -7,33 +7,30 @@ export function sshIssueOverview() {
...
-        </CollapsibleWellComponent>
-      </>
-    );
-  }
+  return (
+    <>
+      Protect private keys with a pass phrase.
+      <CollapsibleWellComponent>
+        The machine <span className="badge badge-primary">{issue.machine}</span> (<span
+          className="badge badge-info" style={{ margin: '2px' }}>{issue.ip_address}</span>) is vulnerable to a <span
```
Comment (shreyamalviya) [inline monkey/monkey_island/cc/ui/src/components/report-components/security/issues/SshIssue.js:29]:
```suggestion
          className="badge badge-info" style={{ margin: '2px' }}>{issue.ip_address}</span>) is vulnerable to an <span
```

File: monkey/monkey_island/cc/services/reporting/report.py [hunk 3]
```diff
@@ -142,39 +148,56 @@ def get_scanners_of_machine(cls, machine: Machine) -> List[Machine]:
...
-        exploiter_info = processor.get_exploit_info_by_dict(exploiter_type, exploit)
-        return exploiter_info
+    @classmethod
+    def process_exploit_event(
+        cls, exploit: ExploitationEvent, password_restored: Dict[ipaddress.IPv4Address, bool]
+    ) -> ExploiterReportInfo:
+        if not cls._machine_repository:
+            raise RuntimeError("Machine repository does not exist")
+
+        target_machine = cls._machine_repository.get_machines_by_ip(exploit.target)[0]
```
Comment (shreyamalviya) [inline monkey/monkey_island/cc/services/reporting/report.py:163]: I don't understand this logic. Why is the first machine always the one we want?

File: monkey/monkey_island/cc/services/reporting/report.py [hunk 4]
```diff
@@ -142,39 +148,56 @@ def get_scanners_of_machine(cls, machine: Machine) -> List[Machine]:
...
+        successful_exploits = filter(lambda x: x.success, exploits)
+        filtered_exploits = ReportService.filter_single_exploit_per_ip(successful_exploits)
+
+        zerologon_events = cls._agent_event_repository.get_events_by_type(PasswordRestorationEvent)
+        password_restored = defaultdict(
+            lambda: None, {e.target: e.success for e in zerologon_events}
+        )
+
+        # Convert the ExploitationEvent into an ExploiterReportInfo
+        return [cls.process_exploit_event(e, password_restored).__dict__ for e in filtered_exploits]
```
Comment (shreyamalviya) [inline monkey/monkey_island/cc/services/reporting/report.py:200]: Use `dataclasses.asdict()` instead of `.__dict__`.

**Review by shreyamalviya — CHANGES_REQUESTED** [review]

File: monkey/monkey_island/cc/services/reporting/report.py (hunk 3 above)
Comment (mssalvatore) [reply monkey/monkey_island/cc/services/reporting/report.py:163]: IPs are not unique, but we're treating them like they are for now. This is an assumption we're making throughout the code until we can come up with a reliable way to disambiguate IP addresses.

**Review by VakarisZ — COMMENTED** [review]
Ideally, the file `services/reporting/report.py` should not exist. I don't think it's necessary to migrate it to UI now, focusing on one thing at a time is fine.

File: monkey/monkey_island/cc/ui/src/components/report-components/security/issues/WmiIssue.js [hunk 5]
```diff
@@ -1,36 +1,17 @@
...
-          The Monkey used a pass-the-hash attack over WMI protocol with user <span
-          className="badge badge-success">{issue.username}</span>.
-        </CollapsibleWellComponent>
-      </>
-    );
-  }
+export function wmiIssueReport(issue) {
+  return (
+    <>
+      Change user passwords to a complex one-use password that is not shared with other computers on the network.
```
Comment (VakarisZ) [inline monkey/monkey_island/cc/ui/src/components/report-components/security/issues/WmiIssue.js:7]: What about the pass the hash?

**Review by VakarisZ — APPROVED** [review]
Haven't found anything worthvile after a quick glance

File: monkey/monkey_island/cc/services/reporting/report.py [hunk 6]
```diff
@@ -142,39 +149,56 @@ def get_scanners_of_machine(cls, machine: Machine) -> List[Machine]:
...
-    @staticmethod
-    def process_exploit(exploit) -> ExploiterReportInfo:
-        exploiter_type = exploit["data"]["exploiter"]
-        exploiter_descriptor = ExploiterDescriptorEnum.get_by_class_name(exploiter_type)
-        processor = exploiter_descriptor.processor()
-        exploiter_info = processor.get_exploit_info_by_dict(exploiter_type, exploit)
-        return exploiter_info
+    @classmethod
+    def process_exploit_event(
+        cls, exploit: ExploitationEvent, password_restored: Dict[ipaddress.IPv4Address, bool]
```
Comment (ilija-lazoroski) [inline monkey/monkey_island/cc/services/reporting/report.py:159]:
```suggestion
        cls, exploitation_event: ExploitationEvent, password_restored: Dict[ipaddress.IPv4Address, bool]
```

File: monkey/monkey_island/cc/services/reporting/report.py [hunk 7]
```diff
@@ -142,39 +149,56 @@ def get_scanners_of_machine(cls, machine: Machine) -> List[Machine]:
...
-            {"$replaceRoot": {"newRoot": "$data"}},
-        ]
-        exploits = []
-        for exploit in mongo.db.telemetry.aggregate(query):
-            new_exploit = ReportService.process_exploit(exploit)
-            if new_exploit not in exploits:
-                exploits.append(new_exploit.__dict__)
-        return exploits
+    def filter_single_exploit_per_ip(
+        exploits: Iterable[ExploitationEvent],
```
Comment (ilija-lazoroski) [inline monkey/monkey_island/cc/services/reporting/report.py:174]:
```suggestion
        exploitation_events: Iterable[ExploitationEvent],
```

**Review by ilija-lazoroski — APPROVED** [review]
Some simple renaming and LGTM!

**Pull request merged**